Update stats plots, add longitudinal sample size calculation

PaulBautin opened the pull request:
This PR intends to homogenize notations and conventions between the graphs presented in the manuscript and the "csa_atrophy" repo. 

Done:
- Update plots: sample_size, error_function_of_csa
- Update legends: boxplot_atrophy, boxplot_csa
- Draw bissection automatically on the boxplot_atrophy graph
- Add graph showing error in function of CSA
- Add longitudinal study sample size computation and update ref for sample size computation
- Update README

FIX #95, FIX #92, FIX #100, FIX #80, FIX #103

PaulBautin committed:
update plots to match manuscript
- update plots: sample_size, error_function_of_csa
- update legends: boxplot_atrophy, boxplot_csa

PaulBautin committed:
add:
- expected trend and mean values on CSA boxplot
- mean values on atrophy boxplot
- add function for adding pearson's r and p-value stats
- add diff and std diff column in dataframe for sample size computation
- add longitudinal study sample size

change:
- correct difference between means (atrophy_% * CSA) on sample size plot
- correct x_label on error_function_of_CSA
- correct x_label on error_function_of_intra_cov
- automatize the detection of outliers on error_function_of_intra_cov_outlier

PaulBautin committed:
- update README for longitudinal study sample size

PaulBautin committed:
update ref in csa_rescale_stat for sample size computation

PaulBautin committed:
correct rescale_estimated_subject in README

jcohenadad committed:
Create output folder if does not exist

PaulBautin committed:
change std to var in sample size

jcohenadad committed:
Added TODOs

PaulBautin committed:
Merge remote-tracking branch 'origin/graph' into graph

# Conflicts:
#	csa_rescale_stat.py

PaulBautin committed:
- correct sample size formula

PaulBautin committed:
- correct diff formula by integrating transformation variability (before this diff was computed from mean CSA across transformations)

PaulBautin committed:
- correct sample size formula

PaulBautin committed:
- remove replacement in diff formula for df_sub

PaulBautin committed:
- use absolute for mean diff and make difference subject dependant

PaulBautin committed:
- remove diff abs because useless (SD uses squared difference anyway)

PaulBautin committed:
- remove diff abs because useless

PaulBautin commented:
Up to now the computed longitudinal sample sizes were small (< 1). After investigation, we observed that the SD of the difference of measured  CSA across subjects were computed using mean CSA across transformations ex:
`diff(sI, rX) = Mean[CSA(sI, r1, :)] - Mean[CSA(sI, rX, :)]` (1)
This had for consequence that the SD of differences did not take into account the variability due to transformations. 


Implemented in 7874439, contrary to formula (1) program does not mean CSA across transformations but randomly samples a CSA value for each subject. ex:
 `diff(sI, rX) = CSA(sI, r1, tY) - CSA(sI, rX, tZ)` (2)
The results with this method are much closer to what was previously found in literature. However, it also puts in evidence a large variability in longitudinal sample sizes due to transformations. This is surprising because the mean intra-subject SD is relatively small (hence, I would not expect such an influence of the transformation-related variability).

jcohenadad commented on csa_rescale_stat.py:
```diff
@@ -412,9 +452,13 @@ def main():
+    df_sub['perc_error'] = 100 * (df_sub['mean'] - df_sub['theoretic_csa']).div(df_sub['theoretic_csa'])
+    diff = []
+    for rescale, group in df.groupby('rescale'):
+        for sub, subgroup in group.groupby('subject'):
```
add comment/explanation

PaulBautin replied:
I've added explanations for the sample size function in commit f50f18a

PaulBautin committed:
- change formula for sample size
- add comments

PaulBautin committed:
- Monte Carlo simulation for between group sample size computation

PaulBautin commented:
With commit 05d7fcd using difference formula (tY and tZ are two different transforms):
> `diff(sI, rX) = CSA(sI, r1, tY) - CSA(sI, rX, tZ)` (2)

longitudinal sample size variability is relatively important

![longitudinal_sample_size](https://user-images.githubusercontent.com/49245398/119000243-f09cca80-b958-11eb-8375-94d5bb3436f5.png)

Also note that, when looking at between-group differences (vs. paired differences as described above), the formula was also updated as follows:

`CSA(sI, rX) = CSA(sI, rX, tZ)` and not `CSA(sI, rX) = MEAN[CSA(sI, rX, :)]`

Results seem to vary much less: SD of sample size no more than 3% of sample size (between groups).

Therefore, my best guess is that the important variability found computing longitudinal sample sizes are mostly due to the variability of CSA measures between scalings (which has already been shown in article). 

@jcohenadad, should we continue with these results? My idea is now to keep the Monte Carlo simulations for both sample size computations.

jcohenadad commented:
@PaulBautin This is an interesting investigation but I need more guidance to understand the formula described in https://github.com/sct-pipeline/csa-atrophy/pull/98#issuecomment-845227755. Without the context of the code I cannot advise on what is the most appropriate solution. I suggest we discuss it in a meeting.

PaulBautin committed:
- add comments for sample size function

PaulBautin committed:
- improve comments for formula var and var_diff

PaulBautin committed:
- limit usage of rescale_area to plots

PaulBautin committed:
- remove rounding before boxplots for csa and atrophy
- normalize by the square of rescale for df['Normalized CSA in mm²']
- use poly1d when plotting trends in plots

PaulBautin committed:
- scatter colorbar takes discrete values
- append fake values for pearson and and p_value for rescale =1

PaulBautin committed:
- compute sample size using 500 itterations
- un-comment concatenating csv files

PaulBautin committed:
- update sample size plot to match article
- change iteration number for computing sample size and print message

PaulBautin commented:
@jcohenadad, could you review? I think this PR is ready to be merged (plots in PR match plots in article).

PaulBautin committed:
- remove ceil sur chaque calcul de sample size

PaulBautin commented:
@jcohenadad, could you review? This PR should be merged into master because plots and stats for the article are based on this PR.

jcohenadad commented:
sorry-- realistically i will not have time to review

sandrinebedard committed:
add egg info

sandrinebedard committed:
change sct_deepseg for sct_deepseg_sc

sandrinebedard committed:
change disc file label

sandrinebedard committed:
add fix for qform sfrom and fix typo disc labels

sandrinebedard committed:
change for sct

sandrinebedard committed:
modify for python 3.9 compatibility

sandrinebedard committed:
setup for compute canada

sandrinebedard committed:
rsync disc file

sandrinebedard committed:
rm qc from sct_deepseg

sandrinebedard committed:
fix missing eextension

sandrinebedard committed:
remove extra QC

sandrinebedard committed:
remove one qc report

sandrinebedard committed:
remove all QC reports

sandrinebedard committed:
remove all qc reports

sandrinebedard committed:
add logging

sandrinebedard committed:
Merge branch 'sb/update-for-ca-python3.9' into graph
